FIPS/fip-draft_foundational_fvm.md

```diff
@@ -0,0 +1,1010 @@
+permissionless community-driven innovation on the Filecoin network.
+
+Some of the envisioned benefits include:
+
+- The proliferation of innovative, trustless solutions in user space.
+  - The ability to program behaviors on top of the existing Filecoin primitives
+    confers new degrees of freedom to developers and increases the potential for
+    innovation.
+  - Some examples include: decentralized compute, Data DAOs, Layer 2 networks,
+    cross-chain bridges, alternative storage markets, programmable storage,
+    collateral loan markets, and more.
+- The migration of functionality from system space to user space, thus resulting
+  in a lower dependence on built-in actor evolution, which is largely
+  bottlenecked on the Core Devs and the FIP process.
+- The emergence of an ecosystem of composable and stackable building blocks and
+  primitives exposed through standardised interfaces, enabling the construction
+  of richer, composite, interconnected solutions.
+- Unification of built-in actors across implementations to speed up protocol
+  development.
+- Enabling automatic/governance-driven protocol upgrades.
+- Potential for more areas of the Filecoin protocol to migrate to WASM space
+  (e.g. block validation, fork choice rule, etc.), it becomes possible to deploy
+  protocol changes as WASM modules to all clients, conditional upon on-chain
+  voting.
+
+## FVM specification
+
+This section provides a fundamental definition of the Filecoin Virtual Machine
+(FVM). It is intended to act as a baseline for future FIPs to enhance and
+evolve, as new capabilities, innovations and mechanisms are worked out.
+
+_Writing style_
+
+The FVM is a system under development, and some aspects are still in flux. For
+this reason, this spec combines two styles to provide a well-rounded picture:
+
+- Normative style, for concrete behaviors, rules, interfaces and types that we
+  propose to introduce imminently. _Conveyed by present tense._
+- Directional style: lays out the thinking for upcoming behaviors, rules,
+  interfaces, and types, without mandating concrete technical choices yet.
+  _Conveyed by future tense and/or speculative language._
+
+### Architectural overview
+
+This diagram provides a high-level architectural overview.
+
+![](./fip-draft_foundational_fvm/overview.png)
+
+### Context and VM interface
+
+In the Filecoin blockchain, the virtual machine (VM) is the component in the
+execution layer responsible for applying transactions over a state tree. These
+transactions are encapsulated in messages. There are two kinds of messages:
+
+- Explicit messages are sent by secp256k1 or BLS account actors, appear on-chain
+  when executed, and carry a signature to prove their authenticity. 
+- Implicit messages are generated by the system, do not appear on chain, and
+  they carry no signature. Examples include cron ticks and reward cycles.
+
+A VM is instantiated with these input parameters:
+
+```rust
+let machine = Machine::new(
```

Comment:
This could confuse readers a bit given that the ref-fvm currently distinguishes between the "executor" (the thing that executes messages) and the "machine" (the thing on which messages are executed).

Reply:
I'll add a note.